File: backend/src/test/java/be/lutske/leolegacy/application/service/ExtractionResultValidationTest.java
package be.lutske.leolegacy.application.service;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Unit tests for {@link ExtractionResult} validation logic.
 * These do NOT require Quarkus — they test pure domain logic.
 */
class ExtractionResultValidationTest {

    @Test
    void completeExtractionHasNoMissingFields() {
        var result = new ExtractionResult(
                "Chocolate Mousse", null, null,
                List.of("200 g chocolate", "4 eggs", "50 g sugar"),
                List.of("Melt chocolate", "Separate eggs", "Fold together"),
                null, null, List.of(), null, null, null);

        assertTrue(result.isComplete());
        assertTrue(result.missingFields().isEmpty());
    }

    @Test
    void missingTitleIsDetected() {
        var result = new ExtractionResult(
                null, null, null,
                List.of("200 g chocolate"),
                List.of("Melt chocolate"),
                null, null, List.of(), null, null, null);

        assertFalse(result.isComplete());
        assertEquals(List.of("title"), result.missingFields());
    }

    @Test
    void blankTitleIsTreatedAsMissing() {
        var result = new ExtractionResult(
                "   ", null, null,
                List.of("200 g chocolate"),
                List.of("Melt chocolate"),
                null, null, List.of(), null, null, null);

        assertFalse(result.isComplete());
        assertTrue(result.missingFields().contains("title"));
    }

    @Test
    void missingIngredientsIsDetected() {
        var result = new ExtractionResult(
                "Chocolate Mousse", null, null,
                null,
                List.of("Melt chocolate"),
                null, null, List.of(), null, null, null);

        assertFalse(result.isComplete());
        assertEquals(List.of("ingredients"), result.missingFields());
    }

    @Test
    void emptyIngredientsListIsTreatedAsMissing() {
        var result = new ExtractionResult(
                "Chocolate Mousse", null, null,
                List.of(),
                List.of("Melt chocolate"),
                null, null, List.of(), null, null, null);

        assertFalse(result.isComplete());
        assertTrue(result.missingFields().contains("ingredients"));
    }

    @Test
    void missingStepsIsDetectedAsMissingPreparation() {
        var result = new ExtractionResult(
                "Chocolate Mousse", null, null,
                List.of("200 g chocolate"),
                null,
                null, null, List.of(), null, null, null);

        assertFalse(result.isComplete());
        assertEquals(List.of("preparation"), result.missingFields());
    }

    @Test
    void emptyStepsListIsTreatedAsMissingPreparation() {
        var result = new ExtractionResult(
                "Chocolate Mousse", null, null,
                List.of("200 g chocolate"),
                List.of(),
                null, null, List.of(), null, null, null);

        assertFalse(result.isComplete());
        assertTrue(result.missingFields().contains("preparation"));
    }

    @Test
    void allFieldsMissingReturnsThreeMissingFields() {
        var result = new ExtractionResult();

        assertFalse(result.isComplete());
        assertEquals(List.of("title", "ingredients", "preparation"), result.missingFields());
    }

    @Test
    void warningsArePreservedInResult() {
        var result = new ExtractionResult(
                "Test", null, null,
                List.of("item"),
                List.of("step"),
                null, null,
                List.of("Handwriting partially illegible", "Servings unclear"),
                null, null, null);

        assertTrue(result.isComplete());
        assertEquals(2, result.warnings().size());
        assertEquals("Handwriting partially illegible", result.warnings().get(0));
    }

    @Test
    void providerAndModelMetadataArePreserved() {
        var result = new ExtractionResult(
                "Test", null, null,
                List.of("item"),
                List.of("step"),
                null, null, List.of(), null,
                "openai", "gpt-4o");

        assertEquals("openai", result.provider());
        assertEquals("gpt-4o", result.model());
    }

    @Test
    void optionalFieldsDoNotAffectCompleteness() {
        var result = new ExtractionResult(
                "Test", null, null,
                List.of("item"),
                List.of("step"),
                null, null, List.of(), null, null, null);

        assertTrue(result.isComplete());
        assertTrue(result.missingFields().isEmpty());
    }

    @Test
    void rawModelResponseIsPreserved() {
        String rawJson = """
                {"title":"Test","ingredients":["item"],"steps":["step"]}""";
        var result = new ExtractionResult(
                "Test", null, null,
                List.of("item"),
                List.of("step"),
                null, null, List.of(),
                rawJson, null, null);

        assertEquals(rawJson, result.rawModelResponse());
    }
}

File: backend/src/test/kotlin/be/lutske/leolegacy/application/service/ExtractionResultValidationTest.kt
package be.lutske.leolegacy.application.service

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

/**
 * Unit tests for [ExtractionResult] validation logic.
 * These do NOT require Quarkus — they test pure domain logic.
 */
class ExtractionResultValidationTest {

    @Test
    fun `complete extraction has no missing fields`() {
        val result = ExtractionResult(
            title = "Chocolate Mousse",
            ingredients = listOf("200 g chocolate", "4 eggs", "50 g sugar"),
            steps = listOf("Melt chocolate", "Separate eggs", "Fold together")
        )

        assertTrue(result.isComplete())
        assertTrue(result.missingFields().isEmpty())
    }

    @Test
    fun `missing title is detected`() {
        val result = ExtractionResult(
            title = null,
            ingredients = listOf("200 g chocolate"),
            steps = listOf("Melt chocolate")
        )

        assertFalse(result.isComplete())
        assertEquals(listOf("title"), result.missingFields())
    }

    @Test
    fun `blank title is treated as missing`() {
        val result = ExtractionResult(
            title = "   ",
            ingredients = listOf("200 g chocolate"),
            steps = listOf("Melt chocolate")
        )

        assertFalse(result.isComplete())
        assertTrue(result.missingFields().contains("title"))
    }

    @Test
    fun `missing ingredients is detected`() {
        val result = ExtractionResult(
            title = "Chocolate Mousse",
            ingredients = null,
            steps = listOf("Melt chocolate")
        )

        assertFalse(result.isComplete())
        assertEquals(listOf("ingredients"), result.missingFields())
    }

    @Test
    fun `empty ingredients list is treated as missing`() {
        val result = ExtractionResult(
            title = "Chocolate Mousse",
            ingredients = emptyList(),
            steps = listOf("Melt chocolate")
        )

        assertFalse(result.isComplete())
        assertTrue(result.missingFields().contains("ingredients"))
    }

    @Test
    fun `missing steps is detected as missing preparation`() {
        val result = ExtractionResult(
            title = "Chocolate Mousse",
            ingredients = listOf("200 g chocolate"),
            steps = null
        )

        assertFalse(result.isComplete())
        assertEquals(listOf("preparation"), result.missingFields())
    }

    @Test
    fun `empty steps list is treated as missing preparation`() {
        val result = ExtractionResult(
            title = "Chocolate Mousse",
            ingredients = listOf("200 g chocolate"),
            steps = emptyList()
        )

        assertFalse(result.isComplete())
        assertTrue(result.missingFields().contains("preparation"))
    }

    @Test
    fun `all fields missing returns three missing fields`() {
        val result = ExtractionResult()

        assertFalse(result.isComplete())
        assertEquals(listOf("title", "ingredients", "preparation"), result.missingFields())
    }

    @Test
    fun `warnings are preserved in result`() {
        val result = ExtractionResult(
            title = "Test",
            ingredients = listOf("item"),
            steps = listOf("step"),
            warnings = listOf("Handwriting partially illegible", "Servings unclear")
        )

        assertTrue(result.isComplete())
        assertEquals(2, result.warnings.size)
        assertEquals("Handwriting partially illegible", result.warnings[0])
    }

    @Test
    fun `provider and model metadata are preserved`() {
        val result = ExtractionResult(
            title = "Test",
            ingredients = listOf("item"),
            steps = listOf("step"),
            provider = "openai",
            model = "gpt-4o"
        )

        assertEquals("openai", result.provider)
        assertEquals("gpt-4o", result.model)
    }

    @Test
    fun `optional fields do not affect completeness`() {
        val result = ExtractionResult(
            title = "Test",
            description = null,
            servings = null,
            ingredients = listOf("item"),
            steps = listOf("step"),
            source = null,
            tags = null
        )

        assertTrue(result.isComplete())
        assertTrue(result.missingFields().isEmpty())
    }

    @Test
    fun `rawModelResponse is preserved`() {
        val rawJson = """{"title":"Test","ingredients":["item"],"steps":["step"]}"""
        val result = ExtractionResult(
            title = "Test",
            ingredients = listOf("item"),
            steps = listOf("step"),
            rawModelResponse = rawJson
        )

        assertEquals(rawJson, result.rawModelResponse)
    }
}
